### PR TITLE
A star search@abhishekgupta368

### PR DESCRIPTION
**Pull request by abhishekgupta368**
A* Search Algorithm #1653
Fixes #

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
-  implemented a* search algorithm
#### Languages Used:
- python
#### Files Added:
- A_Star_Search_Algorithm/A_Star_Search_Algorithm.py
> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by asha15** [comment]
Abishek remove the files which are not relavent to this PR

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    """
+        A node class used in A* Pathfinding.
+        parent: it is parent of current node
+        position: it is current position of node in the maze.
+        g: cost from start to current Node
+        h: heuristic based estimated cost for current Node to end Node
+        f: total cost of present node i.e. :  f = g + h
+    """
+    
+    def __init__(self,parent=None,poistion=None):
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep space after comma 
keep space before and after operators

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        self.parent = parent
+        self.position = poistion
+        self.g=0
+        self.f=0
+        self.h=0
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep space before and after equal sign

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        
+    def __eq__(self,other):
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        return self.position == other.position
+
+class FindPath():
+    def __init__(self,maze,cost,start,end):
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        self.maze = maze
+        self.cost = cost
+        self.start = start
+        self.end = end
+        self.move = [[-1, 0], # go up
+                     [ 0,-1], # go left
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                     [ 1, 0], # go down
+                     [ 0, 1], # go right
+                     [-1,-1], # go left-up
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+                     [-1, 1], # go left down
+                     [ 1,-1], # go right down
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                     [ 1, 1]] # go right up
+        
+    def return_path(self,curr_node,cost_matrix):
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep a space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+            for index, item in enumerate(queue):
+                if item.f < current_node.f:
+                    current_node = item
+                    current_index = index
+                    
+            # if we hit this point return the path such as it may be no solution or 
+            # computation cost is too high
+            if counter > max_steps:
+                print ("Destination cannot be reached")
+                return self.return_path(current_node,self.maze)
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep a space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    
+            # Pop current node out off
+            queue.pop(current_index)
+            # mark it visited
+            visited_list.append(current_node)
+    
+            # check if goal is reached or not
+            if current_node == end_node:
+                return self.return_path(current_node,self.maze)
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep a  space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+            # Generate coordinate from all adjacent coordinates
+            coordinates = []
+    
+            for move in self.move: 
+    
+                # Get node position
+                current_node_position = (current_node.position[0] + move[0], current_node.position[1] + move[1])
+    
+                # check if all the moves are in maze limit
+                if (current_node_position[0] > (no_rows - 1) or current_node_position[0] < 0 or current_node_position[1] > (no_columns -1) or current_node_position[1] < 0):
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: break this into another line

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+                
+                # Child is on the visited list (search entire visited list)
+                if len([visited_child for visited_child in visited_list if visited_child == child]) > 0:
+                    continue
+    
+                # calculate f, g, and h values
+                child.g = current_node.g + self.cost
+                
+                # calculated Heuristic costs, this is using eucledian distance
+                child.h = (((child.position[0] - end_node.position[0]) ** 2) + ((child.position[1] - end_node.position[1]) ** 2))
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: break this into another line

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        self.n=n
+        self.m=m
+    
+    def check(self,value):
+        data=''
+        for i in range(len(value)):
+            if(value[i]=='[' or value[i]==']'):
+                continue
+            else:
+                data+=value[i]
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep a space after comma
keep a space before and after operators

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+        c=0
+        matrix = self.maze
+        matrix = matrix.split(',')
+        data = []
+        for i in range(self.n):
+            l=[]
+            for j in range(self.m):
+                
+                l.append(int(self.check(matrix[c])))
+                c+=1
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep a space after comma
keep a space before and after operators

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+            data.append(l)
+        return data
+
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: remove this extra line

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+                c+=1
+            data.append(l)
+        return data
+        
+        
+if __name__ == '__main__':
+    
+    no_rows = int(input("Enter number of rows: "))
+    no_cols = int(input("Enter number of columns: "))
+    matrix  = Preprocess(str(input("Enter Matrix: ")),no_rows,no_cols).process_text()
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep a space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+    start_x = int(input("Enter x coordinate of starting node: "))
+    start_y = int(input("Enter y coordinate of starting node: "))
+    end_x = int(input("Enter x coordinate of ending node: "))
+    end_y = int(input("Enter y coordinate of ending node: "))
+    cost = int(input("Enter cost: "))
+    start = [start_x,start_y]
+    end = [end_x,end_y]
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep a space after comma

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+    path = FindPath(matrix,cost, start, end).search()
+    if(path!=None):
+        print("Path found: ")
+        for i in range(len(path)):
+            for j in range(len(path[i])):
+                if(path[i][j]== -1):
+                    print(0,end=" ")
+                else:
+                    print(path[i][j],end=" ")
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: keep a space after comma
keep a space before and after operators

File: A_Star_Search_Algorithm/A_Star_Search_Algorithm.py [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+#    Enter x coordinate of starting node: 0
+#    
+#    Enter y coordinate of starting node: 0
+#    
+#    Enter x coordinate of ending node: 4
+#    
+#    Enter y coordinate of ending node: 5
+#    
+#    Enter cost: 1
+#
```
Comment (asha15) [inline A_Star_Search_Algorithm/A_Star_Search_Algorithm.py]: remove the extra spaces

**Comment by abhishekgupta368** [comment]
can you please merge my request, so that, I can work ahead

**Comment by asha15** [comment]
It seems you didn't do the changes i requested. if you did the changes ,please click the resolved button

**Comment by abhishekgupta368** [comment]
i removed space

**Comment by asha15** [comment]
you didn't keep space after comma. pleas reach the code thoroughly and let me knoe

**Comment by abhishekgupta368** [comment]
now, I have checked every thing?

**Comment by jainaman224** [comment]
Fix PR title.

**Comment by jainaman224** [comment]
Please don't comment on every issue. Comment on 2-3 issues at a time.

**Pull request merged**